[quant][graphmode][fx] Refactor node_name_to_target_dtype to make it clearer

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68317

Summary:
We use the node_name_to_target_dtype to store the target dtype for output activations for each node, computed from qconfig for the node,
there are two problems with node_name_to_target_dtype that makes it hard to work with:
1. we mutate node_name_to_target_dtype when we insert observers, this makes the data structure confusing because it's typically unexpected
to change a data structure that store the "target" dtype
2. currently it only stores target dtype about output activations, while we also need target dtype for input activation, weight and bias

This PR fixes both problem by removing mutation from the node_name_to_target_dtype and expanding the target_dtype for node to include
the missing target dtype for input activation, weight and bias. We will have another refactor to simplify the observation for weight and bias dtype
in the future.

Please see comments for the updated structure of node_name_to_target_dtype

TODO: we may want to rename node_name_to_target_dtype to node_name_to_target_dtype_info in a separate PR.

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32411858](https://our.internmc.facebook.com/intern/diff/D32411858)

jerryzh168 committed:
[quant][graphmode][fx] Refactor node_name_to_target_dtype to make it more clear

Summary:
We use the node_name_to_target_dtype to store the target dtype for output activations for each node, computed from qconfig for the node,
there are two problems with node_name_to_target_dtype that makes it hard to work with:
1. we mutate node_name_to_target_dtype when we insert observers, this makes the data structure confusing because it's typically unexpected
to change a data structure that store the "target" dtype
2. currently it only stores target dtype about output activations, while we also need target dtype for input activation, weight and bias

This PR fixes both problem by removing mutation from the node_name_to_target_dtype and expanding the target_dtype for node to include
the missing target dtype for input activation, weight and bias. We will have another refactor to simplify the observation for weight and bias dtype
in the future.

Please see comments for the updated structure of node_name_to_target_dtype

TODO: we may want to rename node_name_to_target_dtype to node_name_to_target_dtype_info in a separate PR.

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32411858).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Refactor node_name_to_target_dtype to make it clearer"


Summary:
We use the node_name_to_target_dtype to store the target dtype for output activations for each node, computed from qconfig for the node,
there are two problems with node_name_to_target_dtype that makes it hard to work with:
1. we mutate node_name_to_target_dtype when we insert observers, this makes the data structure confusing because it's typically unexpected
to change a data structure that store the "target" dtype
2. currently it only stores target dtype about output activations, while we also need target dtype for input activation, weight and bias

This PR fixes both problem by removing mutation from the node_name_to_target_dtype and expanding the target_dtype for node to include
the missing target dtype for input activation, weight and bias. We will have another refactor to simplify the observation for weight and bias dtype
in the future.

Please see comments for the updated structure of node_name_to_target_dtype

TODO: we may want to rename node_name_to_target_dtype to node_name_to_target_dtype_info in a separate PR.

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32411858](https://our.internmc.facebook.com/intern/diff/D32411858)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Refactor node_name_to_target_dtype to make it clearer"


Summary:
We use the node_name_to_target_dtype to store the target dtype for output activations for each node, computed from qconfig for the node,
there are two problems with node_name_to_target_dtype that makes it hard to work with:
1. we mutate node_name_to_target_dtype when we insert observers, this makes the data structure confusing because it's typically unexpected
to change a data structure that store the "target" dtype
2. currently it only stores target dtype about output activations, while we also need target dtype for input activation, weight and bias

This PR fixes both problem by removing mutation from the node_name_to_target_dtype and expanding the target_dtype for node to include
the missing target dtype for input activation, weight and bias. We will have another refactor to simplify the observation for weight and bias dtype
in the future.

Please see comments for the updated structure of node_name_to_target_dtype

TODO: we may want to rename node_name_to_target_dtype to node_name_to_target_dtype_info in a separate PR.

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32411858](https://our.internmc.facebook.com/intern/diff/D32411858)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32411858).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Refactor node_name_to_target_dtype to make it clearer"


Summary:
We use the node_name_to_target_dtype to store the target dtype for output activations for each node, computed from qconfig for the node,
there are two problems with node_name_to_target_dtype that makes it hard to work with:
1. we mutate node_name_to_target_dtype when we insert observers, this makes the data structure confusing because it's typically unexpected
to change a data structure that store the "target" dtype
2. currently it only stores target dtype about output activations, while we also need target dtype for input activation, weight and bias

This PR fixes both problem by removing mutation from the node_name_to_target_dtype and expanding the target_dtype for node to include
the missing target dtype for input activation, weight and bias. We will have another refactor to simplify the observation for weight and bias dtype
in the future.

Please see comments for the updated structure of node_name_to_target_dtype

TODO: we may want to rename node_name_to_target_dtype to node_name_to_target_dtype_info in a separate PR.

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32411858](https://our.internmc.facebook.com/intern/diff/D32411858)

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -908,7 +955,29 @@ def insert_observers_for_model(
+    #   }
+    #
+    # TODO: rename this to node_name_to_target_dtype_info
+    node_name_to_target_dtype: Dict[str, Dict[str, Any]] = defaultdict(dict)
```
should we rename this to something like `original_node_name_to_target_dtype_info`, to clarify that only nodes in the original graph are expected to be in this data structure?

vkuzo replied:
should the type be `Dict[str, Dict[str, Optional[torch.dtype]]]`?

jerryzh168 replied:
sounds good, will rename this in next PR.
will fix the type

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -278,58 +281,105 @@ def get_target_activation_dtype_for_node(
+    if is_activation_post_process_node(arg, modules):
+        observed_arg = arg.args[0]
+        assert isinstance(observed_arg, Node), "Currently we only support observing Node"
+        return node_name_to_target_dtype[observed_arg.name]["output_activation_dtype"]
```
are there any potential cases where we need to skip more than one observer?  I can't think of any, just wondering.

jerryzh168 replied:
No, we don't need to skip multiple observers, there can only be at most one observer based on our implementation the reason is that for each of the "Value" in the graph, we only insert observer when we first create the Value, and then when we use the value (input for some node), for the first case we will have 0 observers, for the second one we have at most 1 observer

vkuzo reviewed (approved):
looks great!

jerryzh168 committed:
Update on "[quant][graphmode][fx] Refactor node_name_to_target_dtype to make it clearer"


Summary:
We use the node_name_to_target_dtype to store the target dtype for output activations for each node, computed from qconfig for the node,
there are two problems with node_name_to_target_dtype that makes it hard to work with:
1. we mutate node_name_to_target_dtype when we insert observers, this makes the data structure confusing because it's typically unexpected
to change a data structure that store the "target" dtype
2. currently it only stores target dtype about output activations, while we also need target dtype for input activation, weight and bias

This PR fixes both problem by removing mutation from the node_name_to_target_dtype and expanding the target_dtype for node to include
the missing target dtype for input activation, weight and bias. We will have another refactor to simplify the observation for weight and bias dtype
in the future.

Please see comments for the updated structure of node_name_to_target_dtype

TODO: we may want to rename node_name_to_target_dtype to node_name_to_target_dtype_info in a separate PR.

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32411858](https://our.internmc.facebook.com/intern/diff/D32411858)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32411858).
